js/start.js

```diff
@@ -444,7 +457,40 @@ var loadProfile = function(landingRoute, onboarded) {
               app.simpleMessageModal.remove = () => {
                 throw new Error('This instance of the simpleMessageModal is globally shared ' +
                   'and should not be removed. When you are done with it, please close it.');
-              };              
+              };
+
+              // any lang changes after our app has loaded will need an app re-start to fully take effect
+              user.on('change:language', function(md, lang) {
+                var restartWarning;
+
+                restartWarning = new Dialog({
+                  title: window.polyglot.t('langChangeRestartTitle'),
+                  message: window.polyglot.t('langChangeRestartMessage'),
+                  buttons: [{
+                    text: 'Restart Later',
+                    fragment: 'restart-later'
+                  }, {
+                    text: 'Restart Now',
+                    fragment: 'restart-now'
+                  }]
+                }).on('click-restart-later', () => {
+                  restartWarning.close();
+                }).on('click-restart-now', () => {
+                  location.reload();
+                }).render().open();
+              });
+
+              // If navigating to a page you're already on, we'll hard refresh so
+              // a cached page is not served, which avoids the issue of it looking
+              // like nothing happened.
+              $(document).on('click', '[href^="#"]', (e) => {
```

Comment:
I put it here because this depends on the router and the router is set-up at this point. Also this is more specific because this isn't on all links, but on all links that match the selector `[href^="#"]`. I suppose I could put this code in there with a couple of `if` checks, but I guess that's a matter of personal preference and I feel more comfortable with a separate function handling this case than one giant function conditionally handling different scenarios.

As of now we're not handling the case of it being triggered by JS other than for the short-cut keys (which is also in this PR). I think, for programatic routing, we need to probably handle this on a case by case basis. There were some complications in my attempts to centralize this on some type of `route` or `hashchange` event (mainly because we have a lot of hash changes that don't result in routes, so it's hard to keep track of the true previous route).

The other option is to "hijack" all the different methods of programatic routing, but there are many of them ('Backbone.history.navigate, app.router.navigate, loadUrl, location.href = "#blah", etc...). I'm thinking we need to put some honus on the developer to understand what page they are on and what page they are routing to and what they want the effect to be. If they want a hard refresh, they call `app.router.refresh()`, otherwise they could use one of the other multitude of approaches.